solidity/contracts/MailboxV2.sol

```diff
@@ -0,0 +1,225 @@
+
+    /**
+     * @notice Emitted when a new message is dispatched via Hyperlane
+     * @param messageId The unique message identifier
+     * @param message Raw bytes of message
+     */
+    event Dispatch(bytes32 indexed messageId, bytes message);
+
+    /**
+     * @notice Emitted when a Hyperlane message is delivered
+     * @param messageId The unique message identifier
+     */
+    event Process(bytes32 indexed messageId);
+
+    // ============ Constructor ============
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(uint32 _localDomain) {
+        localDomain = _localDomain;
+    }
+
+    // ============ Initializer ============
+
+    function initialize(address _defaultModule) external initializer {
+        __ReentrancyGuard_init();
+        __Ownable_init();
+        _setDefaultModule(_defaultModule);
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Sets the default ISM for the Mailbox.
+     * @param _module The new default ISM. Must be a contract.
+     */
+    function setDefaultModule(address _module) external onlyOwner {
+        _setDefaultModule(_module);
+    }
+
+    /**
+     * @notice Dispatches a message to the destination domain & recipient.
+     * @param _destinationDomain Domain of destination chain
+     * @param _recipientAddress Address of recipient on destination chain as bytes32
+     * @param _messageBody Raw bytes content of message body
+     * @return The message ID inserted into the Mailbox's merkle tree
+     */
+    function dispatch(
+        uint32 _destinationDomain,
+        bytes32 _recipientAddress,
+        bytes calldata _messageBody
+    ) external override returns (bytes32) {
+        require(_messageBody.length <= MAX_MESSAGE_BODY_BYTES, "msg too long");
+        // Format the message into packed bytes.
+        bytes memory _message = MessageV2.formatMessage(
+            VERSION,
+            count(),
+            localDomain,
+            msg.sender.addressToBytes32(),
+            _destinationDomain,
+            _recipientAddress,
+            _messageBody
+        );
+
+        // Insert the message ID into the merkle tree.
+        bytes32 _id = _message.id();
+        tree.insert(_id);
+        emit Dispatch(_id, _message);
+        return _id;
+    }
+
+    /**
+     * @notice Attempts to deliver `_message` to its recipient. Verifies
+     * `_message` via the recipient's ISM using the provided `_metadata`.
+     * @param _metadata Metadata used by the ISM to verify `_message`.
+     * @param _message Formatted Hyperlane message (refer to Message.sol).
+     */
+    function process(bytes calldata _metadata, bytes calldata _message)
+        external
+        override
+        nonReentrant
+    {
+        // Check that the message was intended for this mailbox.
+        require(_message.version() == VERSION, "!version");
+        require(_message.destination() == localDomain, "!destination");
```

Comment:
I know its not the most gas efficient, but can we have more readable error messages for developers for the ones that they are likely to encounter?

Reply:
Are they really likely to encounter these error messages?

Reply:
Yeah that's a fair point. I guess the only people might be relayer developers, but that's either further out or they will probably know?